integration/v3_grpc_test.go

```diff
@@ -193,3 +194,130 @@ func TestV3DeleteRange(t *testing.T) {
 		clus.Terminate(t)
 	}
 }
+
+// TestV3WatchFromCurrentRevision tests Watch APIs from current revision.
+func TestV3WatchFromCurrentRevision(t *testing.T) {
+	tests := []struct {
+		putKeys        []string
+		watchRequest   *pb.WatchRequest
+		headerRevision int64
+		wresp          *pb.WatchResponse
+	}{
+		// watch the key, matching
+		{
+			[]string{"foo"},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Key: []byte("foo")}},
+			1,
+			&pb.WatchResponse{
+				// TODO: add Header to reflect.DeepEqual and remove headerRevision
+				Events: []*storagepb.Event{
+					{
+						Kv: &storagepb.KeyValue{Key: []byte("foo"), Value: []byte("bar"), CreateRevision: 1, ModRevision: 1, Version: 1},
+					},
+				},
+			},
+		},
+		// watch the key, non-matching
+		{
+			[]string{"foo"},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Key: []byte("helloworld")}},
+			1,
+			&pb.WatchResponse{
+				Events: []*storagepb.Event{},
+			},
+		},
+		// watch the prefix, matching
+		{
+			[]string{"fooLong"},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("foo")}},
+			1,
+			&pb.WatchResponse{
+				Events: []*storagepb.Event{
+					{
+						Kv: &storagepb.KeyValue{Key: []byte("fooLong"), Value: []byte("bar"), CreateRevision: 1, ModRevision: 1, Version: 1},
+					},
+				},
+			},
+		},
+		// watch the prefix, non-matching
+		{
+			[]string{"foo"},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("helloworld")}},
+			1,
+			&pb.WatchResponse{
+				Events: []*storagepb.Event{},
+			},
+		},
+	}
+
+	for i, tt := range tests {
+		clus := newClusterGRPC(t, &clusterConfig{size: 3})
+
+		wAPI := pb.NewWatchClient(clus.RandConn())
+		wStream, err := wAPI.Watch(context.TODO())
+		if err != nil {
+			t.Fatalf("#%d: wAPI.Watch error: %v", i, err)
+		}
+
+		if err := wStream.Send(tt.watchRequest); err != nil {
+			t.Fatalf("#%d: wStream.Send error: %v", i, err)
+		}
+
+		kvc := pb.NewKVClient(clus.RandConn())
+		for _, k := range tt.putKeys {
+			reqput := &pb.PutRequest{Key: []byte(k), Value: []byte("bar")}
+			if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+				t.Fatalf("#%d: couldn't put key (%v)", i, err)
+			}
+		}
+
+		evs := []*storagepb.Event{}
+		createdWatchId, isCreated := 0, false
+
+		cnt := 0
+		for cnt != len(tt.wresp.Events)+1 {
```

Comment:
this is only for the matching case. for non matching case, we should expect just the creation. Also the current test is also incorrect. we did not ensure that there is no more responses coming through the watcher. 

A better way is to put all expected response into wresp and then loop over that slice to match what we receive from the watcher. Then we need to try to receive again to ensure there is no more (do this in another routine and close the stream after a timeout).